core/threaded/scheduler_GEDF_NP.c

```diff
@@ -25,151 +33,34 @@
-  assert(pqueue_size((pqueue_t*)scheduler->executing_reactions) == 0);
-
-  // Loop until it's time to stop or work has been distributed
-  while (true) {
-    if (scheduler->next_reaction_level == (scheduler->max_reaction_level + 1)) {
-      scheduler->next_reaction_level = 0;
-      LF_MUTEX_LOCK(&env->mutex);
-      // Nothing more happening at this tag.
-      LF_PRINT_DEBUG("Scheduler: Advancing tag.");
-      // This worker thread will take charge of advancing tag.
-      if (_lf_sched_advance_tag_locked(scheduler)) {
-        LF_PRINT_DEBUG("Scheduler: Reached stop tag.");
-        _lf_sched_signal_stop(scheduler);
-        LF_MUTEX_UNLOCK(&env->mutex);
-        break;
-      }
-      LF_MUTEX_UNLOCK(&env->mutex);
-    }
+// Data specific to the GEDF scheduler.
+typedef struct custom_scheduler_data_t {
+  pqueue_t* reaction_q;
+  lf_cond_t reaction_q_changed;
+  size_t current_level;
+  bool solo_holds_mutex; // Indicates sole thread holds the mutex.
+} custom_scheduler_data_t;
 
-    if (_lf_sched_distribute_ready_reactions(scheduler) > 0) {
-      _lf_sched_notify_workers(scheduler);
-      break;
-    }
-  }
-}
+/////////////////// Scheduler Private API /////////////////////////
 
 /**
- * @brief Wait until the scheduler assigns work.
- *
- * If the calling worker thread is the last to become idle, it will call on the
- * scheduler to distribute work. Otherwise, it will wait on
- * 'scheduler->semaphore'.
- *
- * @param worker_number The worker number of the worker thread asking for work
- * to be assigned to it.
+ * @brief Mark the calling thread idle and wait for notification of change to the reaction queue.
+ * @param scheduler The scheduler.
+ * @param worker_number The number of the worker thread.
  */
-void _lf_sched_wait_for_work(lf_scheduler_t* scheduler, size_t worker_number) {
-  // Increment the number of idle workers by 1 and check if this is the last
-  // worker thread to become idle.
-  if (((size_t)lf_atomic_add_fetch32((int32_t*)&scheduler->number_of_idle_workers, 1)) ==
-      scheduler->number_of_workers) {
-    // Last thread to go idle
-    LF_PRINT_DEBUG("Scheduler: Worker %zu is the last idle thread.", worker_number);
-    // Call on the scheduler to distribute work or advance tag.
-    _lf_scheduler_try_advance_tag_and_distribute(scheduler);
-  } else {
-    // Not the last thread to become idle.
-    // Wait for work to be released.
-    LF_PRINT_DEBUG("Scheduler: Worker %zu is trying to acquire the scheduling "
-                   "semaphore.",
-                   worker_number);
-    lf_semaphore_acquire(scheduler->semaphore);
-    LF_PRINT_DEBUG("Scheduler: Worker %zu acquired the scheduling semaphore.", worker_number);
-  }
+inline static void wait_for_other_workers_to_finish(lf_scheduler_t* scheduler, int worker_number) {
+  scheduler->number_of_idle_workers++;
+  tracepoint_worker_wait_starts(scheduler->env, worker_number);
+  LF_COND_WAIT(&scheduler->custom_data->reaction_q_changed);
```

Comment:
The doc comment is correct, but I would argue that the name of this procedure is misleading -- it could exit due to a spurious wakeup, or to a change to the event queue other than all workers finishing the current level. If it isn't possible to make this procedure truly return only when the other workers finish, then maybe its name should be updated and its doc comment should emphasize that it does not guard against spurious wakeups.

I do think that the implementation should work correctly as-is, so this comment should be easy to address.